compiler warning: 'not all control paths return a value'

Thanks to @gmobaz for pointing out that building the R package with Visual Studio compilers results in a few compilation warnings (https://github.com/microsoft/LightGBM/issues/3064#issuecomment-626364626). One of these is 'not all control paths return a value'.

We should eliminate this warning if possible, to save users time in debugging.

## How you are using LightGBM?

LightGBM component: R package

## Environment info

Operating System: Windows 10

CPU/GPU model: N/A

C++ compiler version: Visual Studio 16 2019, MSVC 14.24.28314

CMake version: 3.16.5

Java version: N/A

Python version: N/A

R version: 3.6.1

## Error message and / or logs

> C:\Users\James\AppData\Local\Temp\RtmpKw7PIG\R.INSTALL3afce51189c\lightgbm\src\src\boosting\prediction_early_stop.cpp(86): warning C4715: 'LightGBM::CreatePredictionEarlyStopInstance': not all control paths return a value [C:\Users\James\AppData\Local\Temp\RtmpKw7PIG\R.INSTALL3afce51189c\lightgbm\rc\build\_lightgbm.vcxproj]

![image](https://user-images.githubusercontent.com/7608904/81514250-eb5c7700-9325-11ea-8332-5cd6b55f86b4.png)


## Reproducible example(s)

```shell
Rscript build_r.R
```

## Steps to reproduce

1. Make sure Visual Studio 16 2019 is installed
2. Make sure `Rtools35` is installed and its MinGW is on `PATH`
3. run `Rscript build_r.R` from the root of the repo

--- a/src/boosting/prediction_early_stop.cpp
+++ b/src/boosting/prediction_early_stop.cpp
@@ -74,18 +74,18 @@ PredictionEarlyStopInstance CreateBinary(const PredictionEarlyStopConfig& config
 
 PredictionEarlyStopInstance CreatePredictionEarlyStopInstance(const std::string& type,
                                                               const PredictionEarlyStopConfig& config) {
-  if (type != "none" && type != "multiclass" && type != "binary") {
-    Log::Fatal("Unknown early stopping type: %s", type.c_str());
-  }
-
   if (type == "none") {
     return CreateNone(config);
   } else if (type == "multiclass") {
     return CreateMulticlass(config);
+  } else if (type == "binary") {
+    return CreateBinary(config);
+  } else {
+    Log::Fatal("Unknown early stopping type: %s", type.c_str());
   }
 
-  // "binary"
-  return CreateBinary(config);
+  // Fix for compiler warnings about reaching end of control
+  return CreateNone(config);
 }
 
 }  // namespace LightGBM

--- a/src/boosting/prediction_early_stop.cpp
+++ b/src/boosting/prediction_early_stop.cpp
@@ -74,15 +74,19 @@ PredictionEarlyStopInstance CreateBinary(const PredictionEarlyStopConfig& config
 
 PredictionEarlyStopInstance CreatePredictionEarlyStopInstance(const std::string& type,
                                                               const PredictionEarlyStopConfig& config) {
+
+  if (type != "none" && type != "multiclass" && type != "binary") {
+    Log::Fatal("Unknown early stopping type: %s", type.c_str());
+  }
+
   if (type == "none") {
     return CreateNone(config);
   } else if (type == "multiclass") {
     return CreateMulticlass(config);
-  } else if (type == "binary") {
-    return CreateBinary(config);
-  } else {
-    Log::Fatal("Unknown early stopping type: %s", type.c_str());
   }
+
+  // "binary"
+  return CreateBinary(config);
 }
 
 }  // namespace LightGBM

--- a/src/boosting/prediction_early_stop.cpp
+++ b/src/boosting/prediction_early_stop.cpp
@@ -74,7 +74,6 @@ PredictionEarlyStopInstance CreateBinary(const PredictionEarlyStopConfig& config
 
 PredictionEarlyStopInstance CreatePredictionEarlyStopInstance(const std::string& type,
                                                               const PredictionEarlyStopConfig& config) {
-
   if (type != "none" && type != "multiclass" && type != "binary") {
     Log::Fatal("Unknown early stopping type: %s", type.c_str());
   }

